qubeswhonix/__init__.py

```diff
@@ -18,92 +18,185 @@
 # You should have received a copy of the GNU General Public License along
 # with this program; if not, see <http://www.gnu.org/licenses/>.
 
-'''qubes-core-admin extension for handling Whonix related settings'''
+"""qubes-core-admin extension for handling Whonix related settings"""
 
 import qubes.ext
 import qubes.vm.templatevm
 
+
 class QubesWhonixExtension(qubes.ext.Extension):
-    '''qubes-core-admin extension for handling Whonix related settings'''
+    """qubes-core-admin extension for handling Whonix related settings"""
+
     @staticmethod
     def set_ws_netvm(app, vm):
-        '''Set the default NetVM for a Whonix-Workstation qube.'''
+        """Set the default NetVM for a Whonix-Workstation qube."""
         if isinstance(vm, qubes.vm.templatevm.TemplateVM):
             return
-        template = getattr(vm, 'template', None)
-        # look for appropriate whonix-gateway
-        if (
-            template is not None
-            and 'whonix-default-gw' in template.features
+
+        curr_netvm = getattr(vm, "netvm", None)
+
+        if not curr_netvm:
+            # If VM has no NetVM, assume user doesn't want one.
+            netvm = None
+
+        elif curr_netvm and curr_netvm.features.check_with_template(
+            "whonix-gw", None
         ):
-            netvm = template.features['whonix-default-gw']
+            # If VM's NetVM is a Gateway, use it.
+            netvm = curr_netvm
+
         else:
-            netvm = 'sys-whonix'
-        if netvm in app.domains:
+            feature = vm.features.check_with_template("whonix-default-gw", None)
+            if (
+                feature
+                and feature in app.domains
+                and feature.features.get("whonix-gw", None)
```

Comment:
`feature` is a string, not a virtual machine, so I'm pretty sure this will fail as written. Possible fix (untested):


```suggestion
                feature
                and feature in app.domains
                and app.domains[feature].features.get("whonix-gw", None)
```